Knex imports sqlite in migrate:make when told to use pg

# Environment

Knex version: 1.0.5
Database + version: Postgres 14
OS: macOS 12.3

# Bug

`knex migrate:make foo` wants (and fails) to load `@vscode/sqlite3` even though the client is set to `pg`:

```
❯ yarn knex migrate:make foo
yarn run v1.22.17
$ /Users/semenov/tmp/p1/node_modules/.bin/knex migrate:make foo
Knex: run
$ npm install @vscode/sqlite3 --save
Cannot find module '@vscode/sqlite3'
Require stack:
- /Users/semenov/tmp/p1/node_modules/knex/lib/dialects/sqlite3/index.js
- /Users/semenov/tmp/p1/node_modules/knex/lib/knex-builder/internal/config-resolver.js
- /Users/semenov/tmp/p1/node_modules/knex/lib/knex-builder/Knex.js
- /Users/semenov/tmp/p1/node_modules/knex/lib/index.js
- /Users/semenov/tmp/p1/node_modules/knex/knex.js
- /Users/semenov/tmp/p1/node_modules/knex/bin/cli.js
Error: Cannot find module '@vscode/sqlite3'
Require stack:
- /Users/semenov/tmp/p1/node_modules/knex/lib/dialects/sqlite3/index.js
- /Users/semenov/tmp/p1/node_modules/knex/lib/knex-builder/internal/config-resolver.js
- /Users/semenov/tmp/p1/node_modules/knex/lib/knex-builder/Knex.js
- /Users/semenov/tmp/p1/node_modules/knex/lib/index.js
- /Users/semenov/tmp/p1/node_modules/knex/knex.js
- /Users/semenov/tmp/p1/node_modules/knex/bin/cli.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Client_SQLite3._driver (/Users/semenov/tmp/p1/node_modules/knex/lib/dialects/sqlite3/index.js:33:12)
    at Client_SQLite3.initializeDriver (/Users/semenov/tmp/p1/node_modules/knex/lib/client.js:190:26)
    at new Client (/Users/semenov/tmp/p1/node_modules/knex/lib/client.js:75:12)
    at new Client_SQLite3 (/Users/semenov/tmp/p1/node_modules/knex/lib/dialects/sqlite3/index.js:22:5)
    at knex (/Users/semenov/tmp/p1/node_modules/knex/lib/knex-builder/Knex.js:12:28)
    at initKnex (/Users/semenov/tmp/p1/node_modules/knex/bin/cli.js:74:10)
/Users/semenov/tmp/p1/node_modules/knex/lib/client.js:197
      throw new Error(`${message}\n${e.message}`);
            ^

Error: Knex: run
$ npm install @vscode/sqlite3 --save
Cannot find module '@vscode/sqlite3'
Require stack:
- /Users/semenov/tmp/p1/node_modules/knex/lib/dialects/sqlite3/index.js
- /Users/semenov/tmp/p1/node_modules/knex/lib/knex-builder/internal/config-resolver.js
- /Users/semenov/tmp/p1/node_modules/knex/lib/knex-builder/Knex.js
- /Users/semenov/tmp/p1/node_modules/knex/lib/index.js
- /Users/semenov/tmp/p1/node_modules/knex/knex.js
- /Users/semenov/tmp/p1/node_modules/knex/bin/cli.js
    at Client_SQLite3.initializeDriver (/Users/semenov/tmp/p1/node_modules/knex/lib/client.js:197:13)
    at new Client (/Users/semenov/tmp/p1/node_modules/knex/lib/client.js:75:12)
    at new Client_SQLite3 (/Users/semenov/tmp/p1/node_modules/knex/lib/dialects/sqlite3/index.js:22:5)
    at knex (/Users/semenov/tmp/p1/node_modules/knex/lib/knex-builder/Knex.js:12:28)
    at initKnex (/Users/semenov/tmp/p1/node_modules/knex/bin/cli.js:74:10)
    at async Command.<anonymous> (/Users/semenov/tmp/p1/node_modules/knex/bin/cli.js:213:24)
```

## Reduced test code

In an empty directory:

```
yarn add knex
```

create `knexfile.js`:

```js
module.exports = {
  client: 'pg',
  connection: 'postgres://localhost/test'
}
```

run:

```
pnpm knex migrate:make foo
```

## Cause

This is caused by: https://github.com/knex/knex/blob/f380e48e30663d6ae571f15237fac7fe7d08cb73/bin/cli.js#L212

This fake client option which user never provided overrides whatever `client` is set in config.

The [upgrade guide](https://github.com/knex/knex/blob/master/UPGRADING.md) (referenced [here](https://github.com/knex/knex/issues/5001#issuecomment-1030430724)) says:

> * If you are using `sqlite3` driver dependency, please replace it with `@vscode/sqlite3` in your `package.json`;

Note that it doesn't say that `@vscode/sqlite3` is now obligatory peer dependency when using migrations. This is a bug, there is no point in having `@vscode/sqlite3`, it's not being imported for any reasonable purpose.

--- a/bin/cli.js
+++ b/bin/cli.js
@@ -42,7 +42,7 @@ async function openKnexfile(configPath) {
   return config;
 }
 
-async function initKnex(env, opts) {
+async function initKnex(env, opts, noClientOverride) {
   checkLocalModule(env);
   if (process.cwd() !== env.cwd) {
     process.chdir(env.cwd);
@@ -54,24 +54,27 @@ async function initKnex(env, opts) {
 
   env.configuration = env.configPath
     ? await openKnexfile(env.configPath)
-    : mkConfigObj(opts);
+    : mkConfigObj(opts, noClientOverride);
 
-  const resolvedConfig = resolveEnvironmentConfig(
+  let resolvedConfig = resolveEnvironmentConfig(
     opts,
     env.configuration,
     env.configPath
   );
 
-  const optionsConfig = parseConfigObj(opts);
-  const config = merge(resolvedConfig, optionsConfig);
+  // In the other case, config is already override in mkConfigObj.
+  if (env.configPath) {
+    const optionsConfig = parseConfigObj(opts, noClientOverride);
+    resolvedConfig = merge(resolvedConfig, optionsConfig);
+  }
 
   // Migrations directory gets defaulted if it is undefined.
-  if (!env.configPath && !config.migrations.directory) {
-    config.migrations.directory = null;
+  if (!env.configPath && !resolvedConfig.migrations.directory) {
+    resolvedConfig.migrations.directory = null;
   }
 
   const knex = require(env.modulePath);
-  return knex(config);
+  return knex(resolvedConfig);
 }
 
 function invoke() {
@@ -122,8 +125,6 @@ function invoke() {
     configPath,
     configuration: null,
   };
-
-  // rechoir.prepare(interpret.jsVariants, configPath, cwd);
 
   let modulePackage = {};
   try {
@@ -210,7 +211,7 @@ function invoke() {
     .action(async (name) => {
       const opts = commander.opts();
       opts.client = opts.client || 'sqlite3'; // We don't really care about client when creating migrations
-      const instance = await initKnex(env, opts);
+      const instance = await initKnex(env, opts, true);
       const ext = getMigrationExtension(env, opts);
       const configOverrides = { extension: ext };
 

--- a/bin/cli.js
+++ b/bin/cli.js
@@ -69,11 +69,7 @@ async function initKnex(env, opts, noClientOverride) {
   }
 
   // Migrations directory gets defaulted if it is undefined.
-  if (
-    !env.configPath &&
-    resolvedConfig.migrations &&
-    !resolvedConfig.migrations.directory
-  ) {
+  if (!env.configPath && !resolvedConfig.migrations.directory) {
     resolvedConfig.migrations.directory = null;
   }
 

--- a/bin/cli.js
+++ b/bin/cli.js
@@ -69,7 +69,11 @@ async function initKnex(env, opts, noClientOverride) {
   }
 
   // Migrations directory gets defaulted if it is undefined.
-  if (!env.configPath && !resolvedConfig.migrations.directory) {
+  if (
+    !env.configPath &&
+    resolvedConfig.migrations &&
+    !resolvedConfig.migrations.directory
+  ) {
     resolvedConfig.migrations.directory = null;
   }
 

--- a/bin/utils/cli-config-utils.js
+++ b/bin/utils/cli-config-utils.js
@@ -7,7 +7,7 @@ const color = require('colorette');
 const argv = require('getopts')(process.argv.slice(2));
 
 function parseConfigObj(opts, noClientOverride) {
-  const config = {};
+  const config = { migrations: {} };
 
   if (opts.client && (!noClientOverride || !config.client)) {
     config.client = opts.client;
@@ -17,15 +17,12 @@ function parseConfigObj(opts, noClientOverride) {
     config.connection = opts.connection;
   }
 
-  if (opts.migrationsDirectory || opts.migrationsTableName) {
-    config.migrations = {};
-    if (opts.migrationsDirectory) {
-      config.migrations.directory = opts.migrationsDirectory;
-    }
+  if (opts.migrationsDirectory) {
+    config.migrations.directory = opts.migrationsDirectory;
+  }
 
-    if (opts.migrationsTableName) {
-      config.migrations.tableName = opts.migrationsTableName;
-    }
+  if (opts.migrationsTableName) {
+    config.migrations.tableName = opts.migrationsTableName;
   }
 
   return config;

--- a/bin/utils/cli-config-utils.js
+++ b/bin/utils/cli-config-utils.js
@@ -6,16 +6,10 @@ const tildify = require('tildify');
 const color = require('colorette');
 const argv = require('getopts')(process.argv.slice(2));
 
-function findCaseInsensitiveProperty(propertyName, object) {
-  return Object.keys(object).find(
-    (key) => key.toLowerCase() === propertyName.toLowerCase()
-  );
-}
+function parseConfigObj(opts, noClientOverride) {
+  const config = {};
 
-function parseConfigObj(opts) {
-  const config = { migrations: {} };
-
-  if (opts.client) {
+  if (opts.client && (!noClientOverride || !config.client)) {
     config.client = opts.client;
   }
 
@@ -23,18 +17,21 @@ function parseConfigObj(opts) {
     config.connection = opts.connection;
   }
 
-  if (opts.migrationsDirectory) {
-    config.migrations.directory = opts.migrationsDirectory;
-  }
+  if (opts.migrationsDirectory || opts.migrationsTableName) {
+    config.migrations = {};
+    if (opts.migrationsDirectory) {
+      config.migrations.directory = opts.migrationsDirectory;
+    }
 
-  if (opts.migrationsTableName) {
-    config.migrations.tableName = opts.migrationsTableName;
+    if (opts.migrationsTableName) {
+      config.migrations.tableName = opts.migrationsTableName;
+    }
   }
 
   return config;
 }
 
-function mkConfigObj(opts) {
+function mkConfigObj(opts, noClientOverride) {
   if (!opts.client) {
     throw new Error(
       `No configuration file found and no commandline connection parameters passed`
@@ -44,7 +41,7 @@ function mkConfigObj(opts) {
   const envName = opts.env || process.env.NODE_ENV || 'development';
   const resolvedClientName = resolveClientNameWithAliases(opts.client);
   const useNullAsDefault = resolvedClientName === 'sqlite3';
-  const parsedConfig = parseConfigObj(opts);
+  const parsedConfig = parseConfigObj(opts, noClientOverride);
 
   return {
     ext: DEFAULT_EXT,

--- a/test/cli/migrate-make.spec.js
+++ b/test/cli/migrate-make.spec.js
@@ -136,6 +136,33 @@ module.exports = {
       );
     });
 
+    it('Create new migration with default knexfile with pg client', () => {
+      fileHelper.registerGlobForCleanup(
+        'test/jake-util/knexfile_migrations/*_somename.js'
+      );
+      fileHelper.createFile(
+        process.cwd() + '/knexfile.js',
+        `
+module.exports = {
+  client: 'pg',
+  connection: {
+    filename: __dirname + '/test/jake-util/test.sqlite3',
+  },
+  migrations: {
+    directory: __dirname + '/test/jake-util/knexfile_migrations',
+  },
+};
+    `,
+        { isPathAbsolute: true }
+      );
+      return execCommand(
+        `node ${KNEX} migrate:make somename --knexpath=../knex.js`,
+        {
+          expectedOutput: 'Created Migration',
+        }
+      );
+    });
+
     it('Create new migration with default ts knexfile', async () => {
       fileHelper.registerGlobForCleanup(
         'test/jake-util/knexfile_migrations/*_somename1.ts'

